plugins/Geant4Output2EDM4hep_DRC.cpp

```diff
@@ -0,0 +1,689 @@
+        else if ( mask.isSet(G4PARTICLE_GEN_OTHER) )         mcp.setGeneratorStatus(9);
+      }
+
+      // Set simulation status
+      mcp.setCreatedInSimulation(         mask.isSet(G4PARTICLE_SIM_CREATED) );
+      mcp.setBackscatter(                 mask.isSet(G4PARTICLE_SIM_BACKSCATTER) );
+      mcp.setVertexIsNotEndpointOfParent( mask.isSet(G4PARTICLE_SIM_PARENT_RADIATED) );
+      mcp.setDecayedInTracker(            mask.isSet(G4PARTICLE_SIM_DECAY_TRACKER) );
+      mcp.setDecayedInCalorimeter(        mask.isSet(G4PARTICLE_SIM_DECAY_CALO) );
+      mcp.setHasLeftDetector(             mask.isSet(G4PARTICLE_SIM_LEFT_DETECTOR) );
+      mcp.setStopped(                     mask.isSet(G4PARTICLE_SIM_STOPPED) );
+      mcp.setOverlay(                     false );
+
+      //fg: if simstatus !=0 we have to set the generator status to 0:
+      if( mcp.isCreatedInSimulation() )
+        mcp.setGeneratorStatus( 0 )  ;
+
+      mcp.setSpin(p->spin);
+      mcp.setColorFlow(p->colorFlow);
+
+      p_ids[id] = cnt++;
+      p_part.push_back(p);
+    }
+
+    // Now establish parent-daughter relationships
+    for(size_t i=0; i < p_ids.size(); ++i)   {
+      const Geant4Particle* p = p_part[i];
+      auto q = m_particles[i];
+
+      for (const auto& idau : p->daughters) {
+        const auto k = p_ids.find(idau);
+        if (k == p_ids.end()) {
+          fatal("+++ Particle %d: FAILED to find daughter with ID:%d",p->id,idau);
+          continue;
+        }
+        int iqdau = (*k).second;
+        auto qdau = m_particles[iqdau];
+        q.addToDaughters(qdau);
+      }
+
+      for (const auto& ipar : p->parents) {
+        if (ipar >= 0) { // A parent ID of -1 means NO parent, because a base of 0 is perfectly legal
+          const auto k = p_ids.find(ipar);
+          if (k == p_ids.end()) {
+            fatal("+++ Particle %d: FAILED to find parent with ID:%d",p->id,ipar);
+            continue;
+          }
+          int iqpar = (*k).second;
+          auto qpar = m_particles[iqpar];
+          q.addToParents(qpar);
+        }
+      }
+    }
+  }
+}
+
+/// Callback to store the Geant4 event
+void Geant4Output2EDM4hep_DRC::saveEvent(OutputContext<G4Event>& ctxt)  {
+  EventParameters* parameters = context()->event().extension<EventParameters>(false);
+  int runNumber(0), eventNumber(0);
+  const int eventNumberOffset(m_eventNumberOffset > 0 ? m_eventNumberOffset : 0);
+  const int runNumberOffset(m_runNumberOffset > 0 ? m_runNumberOffset : 0);
+  double eventWeight{0};
+  // Get event number, run number and parameters from extension ...
+  if ( parameters ) {
+    runNumber = parameters->runNumber() + runNumberOffset;
+    eventNumber = parameters->eventNumber() + eventNumberOffset;
+    parameters->extractParameters(m_frame);
+#if podio_VERSION_MAJOR > 0 || podio_VERSION_MINOR > 16 || podio_VERSION_PATCH > 2
+    // This functionality is only present in podio > 0.16.2
+    eventWeight = m_frame.getParameter<double>("EventWeights");
```

Comment:
`getParameter` returns an optional, to access the variable we have to append `.value()` as in the following:
```
    eventWeight = m_frame.getParameter<double>("EventWeights").value();
    ```

Reply:
If I add .value(), then I get this error when compiling

```
/u/user/swkim/KEY4HEP/for_k4geo_pr_tmp/k4geo_memUpdateVer/k4geo/plugins/Geant4Output2EDM4hep_DRC.cpp: In member function 'virtual void dd4hep::sim::Geant4Output2EDM4hep_DRC::saveEvent(dd4hep::sim::Geant4OutputAction::OutputContext<G4Event>&)':
/u/user/swkim/KEY4HEP/for_k4geo_pr_tmp/k4geo_memUpdateVer/k4geo/plugins/Geant4Output2EDM4hep_DRC.cpp:453:64: error: request for member 'value' in '((dd4hep::sim::Geant4Output2EDM4hep_DRC*)this)->dd4hep::sim::Geant4Output2EDM4hep_DRC::m_frame.podio::Frame::getParameter<double>(std::__cxx11::basic_string<char>(((const char*)"EventWeights"), std::allocator<char>()))', which is of non-class type 'podio::GenericDataReturnType<double>' {aka 'double'}
  453 |     eventWeight = m_frame.getParameter<double>("EventWeights").value();
```

Is this expected?

Reply:
I would say it is not expected. I was using key4hep nightlies for Alma9, and I got no error. What key4hep release (stable, nightlies) and what operative system are you using? 
Maybe @jmcarcell can help here :)

Reply:
I am not entirely sure why you need a copy of the Geant4 output action here. In any case, there have been a few changes to the DD4hep one that should make all of this work transparently with all versions of podio and EDM4hep

- https://github.com/AIDASoft/DD4hep/pull/1264
- https://github.com/AIDASoft/DD4hep/pull/1252
- https://github.com/AIDASoft/DD4hep/pull/1251
- https://github.com/AIDASoft/DD4hep/pull/1198

Reply:
It's probably because the code is being compiled on top of the latest release and not the nightlies. I would recommend compiling with the nightlies first and then doing like https://github.com/AIDASoft/DD4hep/pull/1264 and others if needed to work on the release (or alternatively wait for the next release that will work if it works with the nightlies).

Reply:
I needed this plugin to store the custom hit class (Geant4DRCalorimeter::Hit).
I always sourced the nightly builds before compiling the k4geo (`source /cvmfs/sw-nightlies.hsf.org/key4hep/setup.sh`) but still this happened.
I was working on the CentOS7 machine (local server), and testing on Alma9(lxplus) solved the errors.

Anyway I updated this code following the latest dd4hep master branch `Geant4Output2EDM4hep.cpp` code.